text/3-entrees/storage-api/vec-set.md

```diff
@@ -0,0 +1,145 @@
+	// leverage the binary search which makes this check O(log n).
+	match members.binary_search(&new_member) {
+		// If the search succeeds, the caller is already a member, so just return
+		Ok(_) => Err(Error::<T>::AlreadyMember.into()),
+		// If the search fails, the caller is not a member and we learned the index where
+		// they should be inserted
+		Err(index) => {
+			members.insert(index, new_member.clone());
+			Members::<T>::put(members);
+			Self::deposit_event(RawEvent::MemberAdded(new_member));
+			Ok(())
+		}
+	}
+}
+```
+
+If it turns out that the caller is not already a member, the binary search will fail. In this case
+it still returns the index into the `Vec` at which the member would have been stored had he been
+present. We then use this information to insert the member at the appropriate location, thus
+maintaining a sorted `Vec`.
+
+## Removing a Member
+
+Removing a member is entirely straightforward. We begin by looking for the caller in the list. If
+the caller is not present, there is no work to be done. If the caller is present, the search
+algorithm returns her index, and she can be removed.
+
+```rust, ignore
+fn remove_member(origin) -> DispatchResult {
+	let old_member = ensure_signed(origin)?;
+
+	let mut members = Members::<T>::get();
+
+	// We have to find out where, in the sorted vec the member is, if anywhere.
+	match members.binary_search(&old_member) {
+		// If the search succeeds, the caller is a member, so remove her
+		Ok(index) => {
+			members.remove(index);
+			Members::<T>::put(members);
+			Self::deposit_event(RawEvent::MemberRemoved(old_member));
+			Ok(())
+		},
+		// If the search fails, the caller is not a member, so just return
+		Err(_) => Err(Error::<T>::NotMember.into()),
+	}
+}
+```
+
+## Performance
+
+Now that we have built our set, let's analyze its performance in some common operations.
+
+### Membership Check
+
+In order to check for the presence of an item in a `vec-set`, we make a single storage read, decode
+the entire vector, and perform a binary search.
+
+DB Reads: O(1) Decoding: O(n) Search: O(log n)
+
+### Updating
+
+Updates to the set, such as adding and removing members as we demonstrated, requires first
+performing a membership check. It also requires re-encoding the entire `Vec` and storing it back
+in the database.
+
+DB Writes: O(1) Encoding: O(n)
+
+### Iteration
+
+Iterating over all items in a `vec-set` is achieved by using the `Vec`'s own
+[`iter` method](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.iter). The entire set can,
+again, be read from storage in one go, and each item must be decoded. Finally, the actual processing
+you do on the items will take some time.
+
+DB Reads: O(1) Decoding: O(n) Processing: O(n)
+
+Because accessing the database is a relatively slow operation, only executing a single read while
+iterating the entire list is a big win. If your data will be iterated frequently, you may want a
+`vec-set`.
```

Comment:
```suggestion
Because accessing the database is a relatively slow operation, reading the entire list in a single read is a big win. If you need to iterate over the data frequently, you may want a `vec-set`.
```

The merits and demerits of these data structures is a Big Topic™ that people still write PhD dissertations about; this here is obviously not a complete treatment of the subject. I think you're doing a good job of conveying the key point: the costly part is most likely database IO and if your set is small enough to easily fit in working memory, then `VecSet` is probably a good choice. For applications that need to make frequent changes to the data or are too big to load all-at-once, maps are good.
In classic computing, maps are often advocated for because of the `O(1)` lookup performance, but in Substrate this is not necessarily true because of how data is stored in the database (each node is an expensive db lookup); it's good that the text makes this difference obvious.
@danforbes for some context, I'd expect a DB read to be 10x-100x slower than a memory access. Maybe even more.